src/functor.jl

```diff
@@ -100,18 +172,55 @@ end
 _default_walk(f, ::Nothing, ::Nothing) = nothing
 
 """
-    fmap(f, x; exclude = isleaf, walk = Functors._default_walk)
+    fmap(f, x; exclude = Functors.isleaf, walk = Functors._default_walk)
 
-A structure and type preserving `map` that works for all [`functor`](@ref)s.
+A structure and type preserving `map`.
 
-By default, traverses `x` recursively using [`functor`](@ref)
-and transforms every leaf node identified by `exclude` with `f`.
```

Comment:
It's important to know that `exclude` is used before `functor`, it stops the recursion, rather than just controlling whether `f` is applied. I'm not sure that's what we want, but it's what we have.

Reply:
Yes, this is deliberate because it's the easiest way to control the depth of recursion. Looking back on https://github.com/FluxML/Functors.jl/pull/13, I think it might've been better just to call this `isleaf`: that's literally what it's checking and `exclude` could imply that `f` is _not_ being applied to anything the function predicate says is true.

Reply:
Ok. Maybe all names for such things are confusing... 

What it sounds like it might do is let you tell `fmap(sqrt, (2, Foo(3, sin)))` which leaves to apply `f` to, and which to leave alone. I guess you can test `isleaf` first. Maybe all functionality can be written in terms of any other...

Reply:
It's a little confusing because `f` is also used as a name for not the original user callback, but in essence Functors will try to recurse into anything "left alone" by `isleaf`/`exclude`. It's just that the fallback returns an empty tuple of children, so recursion naturally stops if the value in question can't be unpacked.